// @aws-cdk/aws-iot-actions-alpha/lib/dynamodbv2-put-item-action.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DynamoDBv2PutItemAction = void 0;
const jsiiDeprecationWarnings = require("../.warnings.jsii.js");
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const iam = require("aws-cdk-lib/aws-iam");
const role_1 = require("./private/role");
/**
 * The action to put the record from an MQTT message to the DynamoDB table.
 */
class DynamoDBv2PutItemAction {
    /**
     * @param table the DynamoDB table in which to put the items.
     * @param props Optional properties to not use default
     */
    constructor(table, props = {}) {
        this.table = table;
        try {
            jsiiDeprecationWarnings._aws_cdk_aws_iot_actions_alpha_DynamoDBv2PutItemActionProps(props);
        }
        catch (error) {
            if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                Error.captureStackTrace(error, DynamoDBv2PutItemAction);
            }
            throw error;
        }
        this.role = props.role;
    }
    /**
     * @internal
     */
    _bind(rule) {
        const role = this.role ?? (0, role_1.singletonActionRole)(rule);
        role.addToPrincipalPolicy(new iam.PolicyStatement({
            actions: ['dynamodb:PutItem'],
            resources: [this.table.tableArn],
        }));
        return {
            configuration: {
                dynamoDBv2: {
                    putItem: {
                        tableName: this.table.tableName,
                    },
                    roleArn: role.roleArn,
                },
            },
        };
    }
}
exports.DynamoDBv2PutItemAction = DynamoDBv2PutItemAction;
_a = JSII_RTTI_SYMBOL_1;
DynamoDBv2PutItemAction[_a] = { fqn: "@aws-cdk/aws-iot-actions-alpha.DynamoDBv2PutItemAction", version: "2.130.0-alpha.0" };
