// @aws-cdk/aws-iot-actions-alpha/lib/https-action.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.HttpsAction = void 0;
const jsiiDeprecationWarnings = require("../.warnings.jsii.js");
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const role_1 = require("./private/role");
/**
 * The action to send data from an MQTT message to a web application or service.
 */
class HttpsAction {
    /**
     * @param url The url to which to send post request.
     * @param props Optional properties to not use default.
     */
    constructor(url, props = {}) {
        try {
            jsiiDeprecationWarnings._aws_cdk_aws_iot_actions_alpha_HttpsActionProps(props);
        }
        catch (error) {
            if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                Error.captureStackTrace(error, HttpsAction);
            }
            throw error;
        }
        this.url = url;
        this.confirmationUrl = props.confirmationUrl;
        this.headers = props.headers;
        this.role = props.role;
        this.auth = props.auth;
    }
    /**
     * @internal
     */
    _bind(topicRule) {
        const role = this.role ?? (0, role_1.singletonActionRole)(topicRule);
        const sigV4 = this.auth ? {
            sigv4: {
                roleArn: role.roleArn,
                serviceName: this.auth.serviceName,
                signingRegion: this.auth.signingRegion,
            },
        } : this.auth;
        return {
            configuration: {
                http: {
                    url: this.url,
                    confirmationUrl: this.confirmationUrl,
                    headers: this.headers,
                    auth: sigV4,
                },
            },
        };
    }
}
exports.HttpsAction = HttpsAction;
_a = JSII_RTTI_SYMBOL_1;
HttpsAction[_a] = { fqn: "@aws-cdk/aws-iot-actions-alpha.HttpsAction", version: "2.130.0-alpha.0" };
