// @aws-cdk/aws-iot-actions-alpha/lib/iotevents-put-message-action.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.IotEventsPutMessageAction = void 0;
const jsiiDeprecationWarnings = require("../.warnings.jsii.js");
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const role_1 = require("./private/role");
/**
 * The action to put the message from an MQTT message to the IoT Events input.
 */
class IotEventsPutMessageAction {
    /**
     * @param input The IoT Events input to put messages.
     * @param props Optional properties to not use default
     */
    constructor(input, props = {}) {
        this.input = input;
        try {
            jsiiDeprecationWarnings._aws_cdk_aws_iot_actions_alpha_IotEventsPutMessageActionProps(props);
        }
        catch (error) {
            if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                Error.captureStackTrace(error, IotEventsPutMessageAction);
            }
            throw error;
        }
        this.batchMode = props.batchMode;
        this.messageId = props.messageId;
        this.role = props.role;
        if (this.batchMode && this.messageId) {
            throw new Error('messageId is not allowed when batchMode is true');
        }
    }
    /**
     * @internal
     */
    _bind(rule) {
        const role = this.role ?? (0, role_1.singletonActionRole)(rule);
        this.input.grantWrite(role);
        return {
            configuration: {
                iotEvents: {
                    batchMode: this.batchMode,
                    inputName: this.input.inputName,
                    messageId: this.messageId,
                    roleArn: role.roleArn,
                },
            },
        };
    }
}
exports.IotEventsPutMessageAction = IotEventsPutMessageAction;
_a = JSII_RTTI_SYMBOL_1;
IotEventsPutMessageAction[_a] = { fqn: "@aws-cdk/aws-iot-actions-alpha.IotEventsPutMessageAction", version: "2.130.0-alpha.0" };
