// @aws-cdk/aws-iot-actions-alpha/lib/open-search-action.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.OpenSearchAction = void 0;
const jsiiDeprecationWarnings = require("../.warnings.jsii.js");
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const iam = require("aws-cdk-lib/aws-iam");
const role_1 = require("./private/role");
/**
 * The action to write data to an Amazon OpenSearch Service domain.
 */
class OpenSearchAction {
    constructor(domain, props) {
        this.domain = domain;
        this.props = props;
        try {
            jsiiDeprecationWarnings._aws_cdk_aws_iot_actions_alpha_OpenSearchActionProps(props);
        }
        catch (error) {
            if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                Error.captureStackTrace(error, OpenSearchAction);
            }
            throw error;
        }
    }
    /**
     * @internal
     */
    _bind(rule) {
        const role = this.props.role ?? (0, role_1.singletonActionRole)(rule);
        // According to CloudFormation documentation, we only need 'es:ESHttpPut' permission
        // https://docs.aws.amazon.com/iot/latest/developerguide/opensearch-rule-action.html#opensearch-rule-action-requirements
        role.addToPrincipalPolicy(new iam.PolicyStatement({
            actions: ['es:ESHttpPut'],
            resources: [this.domain.domainArn, `${this.domain.domainArn}/*`],
        }));
        return {
            configuration: {
                openSearch: {
                    endpoint: `https://${this.domain.domainEndpoint}`,
                    id: this.props.id,
                    index: this.props.index,
                    type: this.props.type,
                    roleArn: role.roleArn,
                },
            },
        };
    }
}
exports.OpenSearchAction = OpenSearchAction;
_a = JSII_RTTI_SYMBOL_1;
OpenSearchAction[_a] = { fqn: "@aws-cdk/aws-iot-actions-alpha.OpenSearchAction", version: "2.130.0-alpha.0" };
