// @aws-cdk/aws-iot-actions-alpha/lib/step-functions-state-machine-action.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.StepFunctionsStateMachineAction = void 0;
const jsiiDeprecationWarnings = require("../.warnings.jsii.js");
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const iam = require("aws-cdk-lib/aws-iam");
const role_1 = require("./private/role");
const aws_cdk_lib_1 = require("aws-cdk-lib");
/**
 * The action to put the record from an MQTT message to the Step Functions State Machine.
 */
class StepFunctionsStateMachineAction {
    /**
     * @param stateMachine The Step Functions Start Machine which shoud be executed.
     * @param props Optional properties to not use default
     */
    constructor(stateMachine, props) {
        this.stateMachine = stateMachine;
        try {
            jsiiDeprecationWarnings._aws_cdk_aws_iot_actions_alpha_StepFunctionsStateMachineActionProps(props);
        }
        catch (error) {
            if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                Error.captureStackTrace(error, StepFunctionsStateMachineAction);
            }
            throw error;
        }
        this.executionNamePrefix = props?.executionNamePrefix;
        this.role = props?.role;
    }
    /**
     * @internal
     */
    _bind(rule) {
        const role = this.role ?? (0, role_1.singletonActionRole)(rule);
        const stateMachineName = aws_cdk_lib_1.Stack.of(this.stateMachine).splitArn(this.stateMachine.stateMachineArn, aws_cdk_lib_1.ArnFormat.COLON_RESOURCE_NAME).resourceName;
        if (!stateMachineName) {
            throw new Error(`No state machine name found in ARN: '${this.stateMachine.stateMachineArn}'`);
        }
        role.addToPrincipalPolicy(new iam.PolicyStatement({
            actions: ['states:StartExecution'],
            resources: [this.stateMachine.stateMachineArn],
        }));
        return {
            configuration: {
                stepFunctions: {
                    stateMachineName,
                    executionNamePrefix: this.executionNamePrefix,
                    roleArn: role.roleArn,
                },
            },
        };
    }
}
exports.StepFunctionsStateMachineAction = StepFunctionsStateMachineAction;
_a = JSII_RTTI_SYMBOL_1;
StepFunctionsStateMachineAction[_a] = { fqn: "@aws-cdk/aws-iot-actions-alpha.StepFunctionsStateMachineAction", version: "2.130.0-alpha.0" };
