// @aws-cdk/aws-kinesisfirehose-destinations-alpha/lib/s3-bucket.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.S3Bucket = void 0;
const jsiiDeprecationWarnings = require("../.warnings.jsii.js");
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const iam = require("aws-cdk-lib/aws-iam");
const common_1 = require("./common");
const helpers_1 = require("./private/helpers");
/**
 * An S3 bucket destination for data from a Kinesis Data Firehose delivery stream.
 */
class S3Bucket {
    constructor(bucket, props = {}) {
        this.bucket = bucket;
        this.props = props;
        try {
            jsiiDeprecationWarnings._aws_cdk_aws_kinesisfirehose_destinations_alpha_S3BucketProps(props);
        }
        catch (error) {
            if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                Error.captureStackTrace(error, S3Bucket);
            }
            throw error;
        }
        if (this.props.s3Backup?.mode === common_1.BackupMode.FAILED) {
            throw new Error('S3 destinations do not support BackupMode.FAILED');
        }
    }
    bind(scope, _options) {
        const role = this.props.role ?? new iam.Role(scope, 'S3 Destination Role', {
            assumedBy: new iam.ServicePrincipal('firehose.amazonaws.com'),
        });
        const bucketGrant = this.bucket.grantReadWrite(role);
        const { loggingOptions, dependables: loggingDependables } = (0, helpers_1.createLoggingOptions)(scope, {
            logging: this.props.logging,
            logGroup: this.props.logGroup,
            role,
            streamId: 'S3Destination',
        }) ?? {};
        const { backupConfig, dependables: backupDependables } = (0, helpers_1.createBackupConfig)(scope, role, this.props.s3Backup) ?? {};
        return {
            extendedS3DestinationConfiguration: {
                cloudWatchLoggingOptions: loggingOptions,
                processingConfiguration: (0, helpers_1.createProcessingConfig)(scope, role, this.props.processor),
                roleArn: role.roleArn,
                s3BackupConfiguration: backupConfig,
                s3BackupMode: this.getS3BackupMode(),
                bufferingHints: (0, helpers_1.createBufferingHints)(this.props.bufferingInterval, this.props.bufferingSize),
                bucketArn: this.bucket.bucketArn,
                compressionFormat: this.props.compression?.value,
                encryptionConfiguration: (0, helpers_1.createEncryptionConfig)(role, this.props.encryptionKey),
                errorOutputPrefix: this.props.errorOutputPrefix,
                prefix: this.props.dataOutputPrefix,
            },
            dependables: [bucketGrant, ...(loggingDependables ?? []), ...(backupDependables ?? [])],
        };
    }
    getS3BackupMode() {
        return this.props.s3Backup?.bucket || this.props.s3Backup?.mode === common_1.BackupMode.ALL
            ? 'Enabled'
            : undefined;
    }
}
exports.S3Bucket = S3Bucket;
_a = JSII_RTTI_SYMBOL_1;
S3Bucket[_a] = { fqn: "@aws-cdk/aws-kinesisfirehose-destinations-alpha.S3Bucket", version: "2.130.0-alpha.0" };
